include/counter.h

```diff
@@ -22,35 +23,106 @@
+			  */
+
+	bool absolute; /*!< If true ticks are treated as absolute value, else
+			*   it is relative to the counter reading performed
+			*   during the call.
+			*/
+
+	counter_alarm_callback_t handler; /*!< handler called on alarm
+					   *   (cannot be NULL).
+					   */
+
+	u32_t ticks; /*!< In case of absolute flag is set, maximal value that
+		      *   can be set equals wrap value (counter_get_wrap).
+		      *   Otherwise counter_get_max_relative_alarm() returns
+		      *   maximal value that can be set. If counter is clock
+		      *   driven then ticks can be converted to microseconds
+		      *   (see @ref counter_ticks_to_us). Alternatively,
+		      *   counter implementation may count asynchronous events.
+		      */
+};
+
+typedef void (*counter_wrap_callback_t)(struct device *dev, void *user_data);
+
+/** @brief Structure with generic counter features. */
+struct counter_config_info {
+	u32_t max_wrap; /*!< Maximal (default) wrap value on which counter is
+			 *   reset (cleared or reloaded).
+			 */
+
+	u32_t freq; /*!< Frequency of the source clock if synchronous events
+		     *   are counted.
+		     */
+
+	bool count_up; /*!< Flag indicating direction of the counter. If true
+			*   counter is counting up else counting down.
+			*/
+
+	u8_t channels; /*!< Number of channels that can be used for setting
+			*   alarm, see @ref counter_set_alarm.
+			*/
+};
 
 typedef int (*counter_api_start)(struct device *dev);
 typedef int (*counter_api_stop)(struct device *dev);
 typedef u32_t (*counter_api_read)(struct device *dev);
 typedef int (*counter_api_set_alarm)(struct device *dev,
-				     counter_callback_t callback,
-				     u32_t count, void *user_data);
+				const struct counter_alarm_cfg *alarm_cfg);
+typedef int (*counter_api_disable_alarm)(struct device *dev,
+				const struct counter_alarm_cfg *alarm_cfg);
+typedef int (*counter_api_set_wrap_alarm)(struct device *dev,
+					  u32_t ticks,
+					  counter_wrap_callback_t callback,
+					  void *user_data);
 typedef u32_t (*counter_api_get_pending_int)(struct device *dev);
+typedef u32_t (*counter_api_get_wrap)(struct device *dev);
+typedef u32_t (*counter_api_get_max_relative_alarm)(struct device *dev);
 
 struct counter_driver_api {
 	counter_api_start start;
 	counter_api_stop stop;
 	counter_api_read read;
 	counter_api_set_alarm set_alarm;
+	counter_api_disable_alarm disable_alarm;
+	counter_api_set_wrap_alarm set_wrap_alarm;
 	counter_api_get_pending_int get_pending_int;
+	counter_api_get_wrap get_wrap;
+	counter_api_get_max_relative_alarm get_max_relative_alarm;
 };
 
 /**
  * @brief Start counter device in free running mode.
  *
- * Start the counter device. If the device is a 'countup' counter, the
- * counter initial value is set to zero. If it is a 'countdown' counter,
- * the initial value is set to the maximum value supported by the device.
+ * Start the counter device. The counter initial value is set to zero. Counter
+ * wraps at 0xffffffff.
```

Comment:
Now that the API allows counters to count up or down and it is possible to set wrap values, this description became obsolete.

Reply:
removed that part.